docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,500 @@
+contain `?`).
+
+
+## Subtype constraint generation
+
+Subtype constraint generation is an operation on two type schemas `P` and `Q`
+and a list of type variables `L`, producing a list of subtype
+constraints `C`.
+
+We write this operation as a relation as follows:
+
+```
+P <: Q [L] -> C
+```
+
+where `P` and `Q` are type schemas, `L` is a list of type variables `T0, ...,
+Tn`, and `C` is a list of subtype and supertype constraints on the `Ti`.
+
+This relation can be read as "`P` is a subtype match for `Q` with respect to the
+list of type variables `L` under constraints `C`".
+
+
+By invariant, at any point in constraint generation, only one of `P` and `Q` may
+be a type schema (that is, contain `?`), only one of `P` and `Q` may contain any
+of the `Ti`, and neither may contain both.  That is, constraint generation is a
+relation on type-schema/type pairs and type/type-schema pairs, only the type
+element of which may refer to the `Ti`.
+
+### Notes:
+
+- For convenience, ordering matters in this presentation: where any two clauses
+  overlap syntactically, the first match is preferred.
+- This presentation is assuming appropriate well-formedness conditions on the
+  input types (e.g. non-cyclic class hierarchies)
+
+### Syntactic notes:
+
+- `C0 + C1` is the concatenation of constraint lists `C0` and `C1`.
+
+### Rules
+
+- The unknown type `?` is a subtype match for any type `Q` with no constraints.
+- Any type `P` is a subtype match for the unknown type `?` with no constraints.
+- A type variable `T` in `L` is a subtype match for any type schema `Q`:
+  - Under constraint `T <: Q`.
+- A type schema `Q` is a subtype match for a type variable `T` in `L`:
+  - Under constraint `M <: Q`.
+- Any two equal types `P` and `Q` are subtype matches under no constraints.
+- Any type `P` is a subtype match for `dynamic`, `Object`, or `void` under no
+  constraints.
+- `Null` is a subtype match for any type `Q` under no constraints.
+- `FutureOr<P>` is a subtype match for `FutureOr<Q>` with respect to `L` under
+  constraints `C`:
+  - If `P` is a subtype match for `Q` with respect to `L` under constraints `C`.
+- `FutureOr<P>` is a subtype match for `Q` with respect to `L` under
+constraints `C0 + C1`.
+  - If `Future<P>` is a subtype match for `Q` with respect to `L` under
+    constraints `C0`.
+  - And `P` is a subtype match for `Q` with respect to `L` under constraints
+    `C1`.
+- `P` is a subtype match for `FutureOr<Q>` with respect to `L` under constraints
+  `C`:
+  - If `P` is a subtype match for `Future<Q>` with respect to `L` under
+    constraints `C`.
+  - Or `P` is not a subtype match for `Future<Q>` with respect to `L` under
+    constraints `C`
+    - And `P` is a subtype match for `Q` with respect to `L` under constraints
+      `C`
+- A type variable `T` not in `L` with bound `P` is a subtype match for the same
+type variable `T` with bound `Q` with respect to `L` under constraints `C`:
```

Comment:
Ruh roh, I can't do this right now because kernel doesn't currently have a way to represent a type variable whose bound is different due to type promotion.  Let's talk with @kmillikin tomorrow when I'm in Seattle about how we want to represent this in kernel.